crates/ruff_python_formatter/src/expression/mod.rs

```diff
@@ -99,26 +106,56 @@ impl FormatRule<Expr, PyFormatContext<'_>> for FormatExpr {
             Expr::Slice(expr) => expr.format().fmt(f),
         });
 
-        let saved_level = f.context().node_level();
-        f.context_mut().set_node_level(NodeLevel::Expression);
-
         let result = match parentheses {
-            Parentheses::Always => {
-                write!(
-                    f,
-                    [group(&format_args![
-                        text("("),
-                        soft_block_indent(&format_expr),
-                        text(")")
-                    ])]
-                )
-            }
+            Parentheses::Always => parenthesized("(", &format_expr, ")").fmt(f),
             // Add optional parentheses. Ignore if the item renders parentheses itself.
-            Parentheses::Optional => optional_parentheses(&format_expr).fmt(f),
-            Parentheses::Custom | Parentheses::Never => Format::fmt(&format_expr, f),
-        };
+            Parentheses::Optional => {
+                if can_omit_optional_parentheses(item, f.context()) {
+                    let saved_level = f.context().node_level();
+
+                    let parens_id = f.group_id("optional_parentheses");
+
+                    f.context_mut()
+                        .set_node_level(NodeLevel::Expression(Some(parens_id)));
 
-        f.context_mut().set_node_level(saved_level);
+                    let result = group(&format_args![
+                        if_group_breaks(&text("(")),
+                        indent_if_group_breaks(
+                            &format_args![soft_line_break(), format_expr],
```

Comment:
why is the line break inside the indent?

Reply:
`Indent` is a bit counter-intuitive. Luckily, this is hidden a way by `soft_block_indent` and `block_indent`. `indent` only increments the indentation level, it doesn't insert any line breaks itself. The `soft_line_break` is necessary to then start a new line, with the new, incremented indent. 

This is why the *closing* newline is outside of the indent. We first need to decrement the indent counter, and only then start the new line.